File: apps/jobnest/src/lib/api/applications.ts
import {
  commands,
  type ApplicationHistoryEvent,
  type ApplicationListItem,
  type ApplicationStatusGroup,
  type CreateTrackedApplicationInput,
  type UpdateApplicationStatusInput,
  type UpdateTrackedApplicationInput,
} from "./bindings";
import { executeLocalApiCall } from "./client";

export type {
  ApplicationHistoryEvent,
  ApplicationListItem,
  ApplicationStatusGroup,
  CreateTrackedApplicationInput,
  UpdateApplicationStatusInput,
  UpdateTrackedApplicationInput,
};

export const applicationsApi = {
  list(): Promise<ApplicationStatusGroup[]> {
    return executeLocalApiCall(() => commands.listApplications());
  },

  listHistory(): Promise<ApplicationHistoryEvent[]> {
    return executeLocalApiCall(() => commands.listApplicationHistory());
  },

  create(
    input: CreateTrackedApplicationInput
  ): Promise<ApplicationListItem> {
    return executeLocalApiCall(() => commands.createTrackedApplication(input));
  },

  update(
    input: UpdateTrackedApplicationInput
  ): Promise<ApplicationListItem> {
    return executeLocalApiCall(() => commands.updateTrackedApplication(input));
  },

  updateStatus(input: UpdateApplicationStatusInput): Promise<void> {
    return executeLocalApiCall(async () => {
      await commands.updateApplicationStatus(input);
    });
  },

  remove(applicationId: string): Promise<void> {
    return executeLocalApiCall(async () => {
      await commands.deleteTrackedApplication(applicationId);
    });
  },
};

File: apps/jobnest/src-tauri/src/services/applications.rs
use crate::{
    db::{AppError, Database},
    models::{ApplicationListItem, CreateTrackedApplicationInput, UpdateTrackedApplicationInput},
};

#[derive(Debug, Clone)]
pub struct ApplicationsService {
    db: Database,
}

impl ApplicationsService {
    pub fn new(db: Database) -> Self {
        Self { db }
    }

    pub async fn list(&self) -> Result<Vec<ApplicationListItem>, AppError> {
        self.db.list_applications().await
    }

    pub async fn create_tracked(
        &self,
        input: CreateTrackedApplicationInput,
    ) -> Result<ApplicationListItem, AppError> {
        self.db.create_tracked_application(input).await
    }

    pub async fn update_tracked(
        &self,
        input: UpdateTrackedApplicationInput,
    ) -> Result<ApplicationListItem, AppError> {
        self.db.update_tracked_application(input).await
    }

    pub async fn delete_tracked(&self, application_id: &str) -> Result<(), AppError> {
        self.db.delete_tracked_application(application_id).await
    }
}
